### ISSUE TITLE
Custom cross references with extra user-provided parameters

### ISSUE THREAD

**Issue by MaskRay** [prabirshrestha/vim-lsp#176]
Currently `autoload/lsp/ui/vim.vim` contains quite some boilerplate for similar requests like `function! lsp#ui#vim#type_definition() abort`.

Some language servers provide custom cross reference requests (e.g. `$ccls/vars`). https://github.com/prabirshrestha/vim-lsp/wiki/Servers-ccls

* params: `TextDocumentPositionParams` or its subclass
* result: `Location[] | null`

Here is an example request `$ccls/vars` which has an additional parameter kind

```javascript
{"jsonrpc":"2.0","method":"$ccls/vars","params":{
  "textDocument":{"uri":"file:///home/ray/llvm/include/llvm/IR/PassManager.h"},
  "position":{"line":320,"character":11},
  "kind":2      // additional request parameter
},
"id":1189}
```

`textDocument/references`:

```javascript
{"jsonrpc":"2.0","method":"textDocument/references","params":{
  "textDocument":{"uri":"file:///home/ray/llvm/include/llvm/IR/PassManager.h"},
  "position":{"line":320,"character":11},
  "context":{"includeDeclaration":false}  // additional request parameter
},
"id":1203}
```

This will allow users to do things similar to:

```vim
" bases of up to 3 levels
nn <silent> xB :call LanguageClient#findLocations({'method':'$ccls/inheritance','levels':3})<cr>
" derived of up to 3 levels
nn <silent> xD :call LanguageClient#findLocations({'method':'$ccls/inheritance','derived':v:true,'levels':3})<cr>

" caller
nn <silent> xc :call LanguageClient#findLocations({'method':'$ccls/call'})<cr>
" callee
nn <silent> xC :call LanguageClient#findLocations({'method':'$ccls/call','callee':v:true})<cr>

nn xv :call LanguageClient#findLocations({'method':'$ccls/vars'})<cr>
nn xV :call LanguageClient#findLocations({'method':'$ccls/vars','kind':1})<cr>
nn xx x
```

And we can save some work from adapters like https://github.com/pdavydov108/vim-lsp-cquery @pdavydov108

**Comment by MaskRay** [comment]
Sorry I would have no energy to implement this.. But I would love to see this implemented, for more awareness of custom cross references among language servers/clients (and of course selfishly an advertisement to [ccls](https://github.com/MaskRay/ccls))

**Comment by m-pilia** [comment]
Lol, I just stumbled upon this issue, I wish I saw it before. Recently I implemented [vim-lsp-ccls](https://github.com/m-pilia/vim-lsp-ccls), which leverages the vim-lsp API and implements a UI to navigate call/inheritance/member trees. It seems to work, despite being still immature, and I would like to improve it.

I got stuck on one point. I would like to build the tree lazily, not building the whole tree at once, but fetching the children of a node only when that node is expanded by the user. Basically, a call to the method `$ccls/call` returns a list of direct children for a node (each of them identified by and an id). To get the children of a given child, recursively, I need to make a new call to `$ccls/call`, containing among the parmetres the id of the node of which I want to retrieve the children. Such calls are made when the user expands a node inside the scratch buffer displaying the tree.

This requires being able to call `lsp#send_request()` when the buffer showing the tree (open in a split window) is active. Unfortunately, as far as I understand, a call to `lsp#send_request()` [fails if the current buffer is not open](/prabirshrestha/vim-lsp/blob/master/autoload/lsp.vim#L310), which is the case for the scratch buffer of course, and I do not see obvious workarounds using the public API of vim-lsp (my current workaround is to [jump back to the previous window](/m-pilia/vim-lsp-ccls/blob/master/autoload/ccls/messages.vim#L27), which is a horribly brittle solution of course).

Any suggestion on how this could be approached? I could work on a PR to vim-lsp if necessary.
